### DISCUSSION TITLE
JSCutlery/semver integration

### DISCUSSION THREAD

**Discussion by dianjuar** [bikecoders/ngx-deploy-npm#73] [Ideas]
Let's discuss about integration, know issues, patches needed, approaches, experiences, current successful histories and anything related about the simbiosis between ngx-deploy-npm and [JSCutlery/semver](https://github.com/JSCutlery/semver#readme)

**Comment by spaceribs** [comment]
The RFC at https://github.com/jscutlery/semver/pull/317 continues to use the pattern of pre/post tasks referenced in the configuration, which effectively solves most of the issues I've run into. The only issue I see with this pattern is that now we have one very big task which encapsulates semver bumping, building, packaging and delivering.

**Reply by edbzn** [reply]
The RFC is about scheduling a target post-release and resolving the options for it, releasing and publishing will still be two decoupled tasks.

**Reply by edbzn** [reply]
Here is the related discussion: https://github.com/jscutlery/semver/issues/167, we want to schedule post-release targets to be able to do things like:

- publish to npm (using ngx-deploy-npm)
- publish GitHub release
- ... 

If you think about a better way to trigger post-release targets, I would listen to you your suggestions.

**Comment by edbzn** [comment]
Hi, so it's not a RFC anymore, `postTargets` are merged in @jscutlery/semver. Here is an example of using it: 

```json
{
  "targets": {
    "version": {
      "executor": "@jscutlery/semver:version",
      "options": {
        "postTargets": ["my-project:github"]
      }
    },
    "github": {
      "executor": "@jscutlery/semver:github",
      "options": {
        "tag": "${tag}",
        "notesFile": "libs/my-project/CHANGELOG.md"
      }
    }
  }
}
```

I would like to reference ngx-deploy-npm in the documentation and give a configuration example. @dianjuar Could you help me to do so? The final objective is being able to deploy to npm directly in the CI.

**Reply by dianjuar** [reply]
Yeah for sure. I think the more straightforward case would be:

```json
"npm": {
  "executor": "ngx-deploy-npm:deploy",
  "options": {
    "access": "public"
  }
}
```

And for Angular Libraries the recommended configuration is:

```json
"npm": {
  "executor": "ngx-deploy-npm:deploy",
  "options": {
    "access": "public",
    "buildTarget": "production"
  }
}
```

With that, is going to build the package and then published it to NPM.

Not the expected result? You can search all the options available in the [documentation](https://github.com/bikecoders/ngx-deploy-npm#-options-)

---

@edbzn I didn't test it out, would be nice to perform an actual test to see how it goes. The option `--dry-run` for sure is going to be handy for this case, no publishing to NPM would be performed.

**Reply by edbzn** [reply]
Thanks, I will do a test asap, one thing not precise to me is how to handle the OTP in the CI. Have you ever do that?

**Reply by dianjuar** [reply]
@edbzn I have some questions about that `postTargets`

### How is going to be generated?

By hand or your plugin is going to create it?

If it's going to be created by hand it's important to notice that as far as I know, Angular libraries have a production mode when compiling, so to be able to publish the library to NPM the option `"buildTarget": "production"` needs to be placed.

If your plugin is going to generate it, you should check if the library has a production configuration on the `build` target.

### That `postTargets` is independent (unique for every library on your workspace) or global (every library is going to use that config)

The problem here is that Angular libraries (as far as I know) have a production compilation target and the rest of the libraries (react, node, nest) don't. So ideally, would be nice to have an independent `postTarget` for each library.

---

The `ng-add` of `ngx-deploy-npm` make that check and create it when is modifying the `workspace.json`. You can check it [here](https://github.com/bikecoders/ngx-deploy-npm/blob/master/src/ng-add.ts#L59)

**Reply by edbzn** [reply]
For now it's configurable by hand, but in the future I would like to provide a generator to make it easier. Regarding the mandatory `build` target, for now I think it's ok to just prevent the user in the documentation, and I will rethink that when I'm gonna work on the generator.

The `postTargets` option is defined on the `@jscutlery/semver:version` executor which is either global (sync mode) or independent (per library), it depends on the mode you choose. I think there's pitfall using this option in a global mode and it does not really make sense. Maybe I could provide more documentation here, noted.

Good to know about the `buildTarget` option for `production` configuration.

**Reply by dianjuar** [reply]
> Thanks, I will do a test asap, one thing not precise to me is how to handle the OTP in the CI. Have you ever do that?

I've never come across with the need to use that option. I have two-factor auth in my NPM but I don't use the `OTP` option. I just create a token and follow [these instructions](https://github.com/bikecoders/ngx-deploy-npm#circleci) (there are for CircleCI but you can follow the same principle in other CI's.)

---

I guess if currently, you want to use the `OTP` option on the configuration file (`workspace.json`) you have to write it down on the file (VERY BAD IDEA, it's super dangerous, not recommended at all ☠️) or create a preparation process on your CI to put the OTP value on your workspace file. Currently, the `workspace.json` and `ngx-deploy-npm` neither both support the option to grab values from the environment variables, which I consider would be the best solution.

**Reply by edbzn** [reply]
Ok I will dig into this, [first findings](https://dev.to/bnb/securely-automating-npm-publish-with-the-new-npm-automation-tokens-oei). I'm sure it will be fun. :smile:

**Reply by dianjuar** [reply]
Now that #83 and #84 are merged, is possible to use semver's `postTargets` feature.

PR's are welcome

**Reply by dianjuar** [reply]
Issue to address the integration #85

**Reply by dianjuar** [reply]
Feels pretty cool to solve this integration issue with your own projects 😄

**Reply by dianjuar** [reply]
@edbzn I made my test with our plugins, and everything works as expected. I don't have a new version to deploy yet, so I haven't thoroughly test it.

The final configuration would look like this:

```json
"deploy": {
  "executor": "ngx-deploy-npm:deploy",
  "options": {
    "access": "public"
  }
},
"version": {
  "executor": "@jscutlery/semver:version",
  "options": {
    "postTargets": ["my-project:deploy"],
    "versionTagPrefix": "v",
    "baseBranch": "master"
  }
}
```

Or can take a look at our [workspace.json](https://github.com/bikecoders/ngx-deploy-npm/blob/master/workspace.json#L58-L71) for a real implementation

**Comment by dianjuar** [comment]
@edbzn you mentioned on #70 the following 

> I can already see conflicting things between ngx-deploy-npm and @jscutlery/semver, this option --package-version for example can conflict/override what semver did just before.

I can't see the problem there, let me explain.
That option is completely optional and was designed to replicate the same behavior that Angular Does with their packages, you can [see](https://github.com/angular/angular/blob/master/packages/common/package.json#L3) that they have a placeholder on the version number and another process is going to generate the version number. If you build your lib that place holder is going to end up on the final package.json, with that option you can just overwrite that place holder with the number you want.

There is something that I'm not seeing?

**Reply by edbzn** [reply]
This option is conflicting because if I release my package using semver, it will write down the version in the `package.json`, after that I use ngx-deploy-npm with the option `--package-version`, no reason to do it but Murphy law tells that everything's possible, then it will publish on npm using the specified version instead of the correct one calculated by semver. Am I wrong?

**Reply by dianjuar** [reply]
If you already have the version wanted on the right package.json there is no need to use `--package-version`. Just don't use that option, it is optional and nothing will happen if you don't use it.

**Reply by dianjuar** [reply]
If some users decide to implement `--package-version` along with semver, well I would say that there is no need, semver is handling everything for you, just deploy the thing and that's it.

I think I got your point, you are foreseeing a possible misuse of the option `--package-version` but I think there is nothing to fear.

**Reply by dianjuar** [reply]
Remember that `--package-version` is optional on ngx-deploy-npm, by default is going to publish the version specified on the package.json.

**Reply by dianjuar** [reply]
Please let me know if there is something that I'm not understanding.

**Reply by spaceribs** [reply]
I'm using `--package-version`, and here's how/why:

When users create new pull requests on my system, I use this option to publish a canary of the affected projects. For that action, I do not want to create a commit, and the version number gets overridden to something like `0.0.0-pr-101-82a57c4` and tagged to be downloaded as `@spaceribs/plopdown-file@pr-101`. To be clear, I'm not using semver at all for this.

**Reply by edbzn** [reply]
Interesting @spaceribs, it's another topic but you caught my attention, can you show me the code where you do that if it's not private?

**Reply by spaceribs** [reply]
unfortunately it is in a private repo, but the command I run in the pull request after the build job looks like this:

```bash
nx affected --target="publish" --tag="pr-${CIRCLE_PR_NUMBER}" --package-version="0.0.0-pr-${CIRCLE_PR_NUMBER}-${CIRCLE_BUILD_NUM}"
```

**Comment by dianjuar** [comment]
@edbzn Now that this tool is fully migrated to use NxDevkit, you can use it to deploy Semver.

As Semver uses Semver to generate Semver's changelogs, you can test the integration with your project.
I'm doing the same with this project now that here we are using Semver (#83)

I will document the process right here 👇

**Reply by edbzn** [reply]
Hey, we made it yesterday for [@jscutlery/devkit](https://github.com/jscutlery/devkit), and it's fully working in the CI.

**Reply by dianjuar** [reply]
You can share your experience here on https://github.com/bikecoders/ngx-deploy-npm/discussions/81 🎉

**Reply by dianjuar** [reply]
I did the integration in this plugin as well. Waiting to have a new version to publish to fully test it out

**Reply by dianjuar** [reply]
@edbzn I had to update Semver to the last version to trigger postTargets properly, see #93

**Comment by dianjuar** [comment]
@edbzn, do you have anything more to discuss, clarify or test about the integration?

For my part, I'm more than satisfied 😁.
Everything is running as expected. I'm using the integration here in the project, and your plugin is also using it, which does two big projects implementing the integration successfully. 

All the tasks related to the integration are complete, and we have already documented the process.

I want to mark this discussion as complete unless you have something more to bring to the table. (I don't know how to complete it 😅)

**Reply by edbzn** [reply]
@dianjuar For my part, it's working like a charm, I think @jscutlery/semver and ngx-depy-npm work seamlessly together. One thing I plan to do soon is canary releases as your package allows it.

I think you cannot mark the discussion as complete as it's not a Q/A.

**Reply by dianjuar** [reply]
@edbzn you're right about the discussion.

The only thing that I can do is lock the conversation to "complete it".

@edbzn @spaceribs Thank you for everything.
It was fun ❤️
